File: src/ARCtrl/Template.Web.fs
module ARCtrl.Template.Web

open ARCtrl
open Fable.Core


let getTemplates(url: string option) =
    let defaultURL = @"https://github.com/nfdi4plants/Swate-templates/releases/download/latest/templates_v2.0.0.json"

    let url = defaultArg url defaultURL
    async {
        let! jsonString = ARCtrl.WebRequest.downloadFile url
        let mapResult = Json.Templates.fromJsonString jsonString
        return mapResult
    }

#if FABLE_COMPILER_JAVASCRIPT

/// <summary>
/// This class is used to make async functions more accessible from JavaScript.
/// </summary>
[<AttachMembers>]
type WebController =
    static member getTemplates(url: string option) =
        async {
            let! templates = getTemplates(url)
            return templates
        }
        |> Async.StartAsPromise
#endif

File: js/Template.Web.js
import { defaultArg } from "./fable_modules/fable-library-js.4.22.0/Option.js";
import { singleton } from "./fable_modules/fable-library-js.4.22.0/AsyncBuilder.js";
import { downloadFile } from "./WebRequest/WebRequest.js";
import { Templates_fromJsonString } from "./JsonIO/Table/Templates.js";
import { startAsPromise } from "./fable_modules/fable-library-js.4.22.0/Async.js";
import { class_type } from "./fable_modules/fable-library-js.4.22.0/Reflection.js";

export function getTemplates(url) {
    const url_1 = defaultArg(url, "https://github.com/nfdi4plants/Swate-templates/releases/download/latest/templates_v2.0.0.json");
    return singleton.Delay(() => singleton.Bind(downloadFile(url_1), (_arg) => {
        const mapResult = Templates_fromJsonString(_arg);
        return singleton.Return(mapResult);
    }));
}

export class WebController {
    constructor() {
    }
    static getTemplates(url) {
        return startAsPromise(singleton.Delay(() => singleton.Bind(getTemplates(url), (_arg) => singleton.Return(_arg))));
    }
}

export function WebController_$reflection() {
    return class_type("ARCtrl.Template.Web.WebController", undefined, WebController);
}

